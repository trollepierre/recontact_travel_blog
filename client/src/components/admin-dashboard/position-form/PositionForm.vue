<template>
  <form class="position__form">
    <p class="position__text">
      {{ $t("lastPosition") }}
    </p>
    <label
      class="position__label"
      for="position-place">
      {{ $t("place") }}
    </label>
    <input
      id="position-place"
      v-model="place"
      class="position__input position__place"
      placeholder="Paris">
    <label
      class="position__label"
      for="position-time">
      {{ $t("time") }}
    </label>
    <input
      id="position-time"
      v-model="time"
      class="position__input position__time"
      placeholder="le 1er mai 2018">
    <button
      type="submit"
      class="position__button position__action--send"
      @click.prevent="updateLastPosition">
      {{ $t("confirm") }}
    </button>
  </form>
</template>
<script>
  import positionsApi from '../../../api/positions'
  import notificationsService from '../../../services/notifications'

  export default {
    name: 'PositionForm',
    data() {
      return {
        place: null,
        time: null,
      }
    },
    methods: {
      submit(e) {
        e.preventDefault()
        this.updateLastPosition()
      },
      updateLastPosition() {
        if (this.place === null || this.time === null) {
          this.displayErrorMessage()
        } else {
          const position = {
            place: this.place,
            time: this.time,
          }
          positionsApi.add(position)
            .then(this.updateLastPositionData)
            .then(this.displaySuccessMessage)
        }
      },
      updateLastPositionData() {
        this.$emit('updateLastPositionData', { place: this.place, time: this.time })
        this.resetPosition()
      },
      resetPosition() {
        this.place = null
        this.time = null
      },
      displayErrorMessage() {
        notificationsService.error(this, this.$t('positionNotUpdated'))
      },
      displaySuccessMessage() {
        notificationsService.success(this, this.$t('positionUpdated'))
      },
    },

    i18n: {
      messages: {
        fr: {
          place: 'Position&nbsp:',
          time: 'Date&nbsp:',
          lastPosition: 'Nouvelle position&nbsp:',
          confirm: 'Envoyer',
          lastKnownPosition: 'Dernière position connue&nbsp:',
          positionUpdated: 'Position mise-à-jour',
          positionNotUpdated: 'Renseigne toutes les informations. La position n\'a pas été mise-à-jour',
        },
        en: {
          place: 'Position:',
          time: 'Date:',
          lastPosition: 'New position:',
          confirm: 'Confirm',
          lastKnownPosition: 'Last known position:',
          positionUpdated: 'Position updated',
          positionNotUpdated: 'Fill all the informations. Position was not updated',
        },
      },
    },
  }
</script>
<style scoped>
  .position__input {
    background: #ffffff;
    border: 1px solid cadetblue;
    padding: 15px 10px;
    border-radius: 4px;
    width: 230px;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .position__button {
    text-transform: uppercase;
    color: #f76252;
    background: #ffffff;
    border: 1px solid #d14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 230px;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .position__button:hover {
    background: #d14800;
    color: #ffffff;
  }

  .position__button:disabled {
    background: #BDBDBD;
    border-color: #616161;
    color: #FAFAFA;
    cursor: auto;
  }

</style>
