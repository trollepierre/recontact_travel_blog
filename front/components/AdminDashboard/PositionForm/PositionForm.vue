<template>
  <form class="position__form">
    <span class="position__language">
      <p class="positionFr__text">
        {{ $t("lastPositionFr") }}
      </p>
      <span class="form-group">
        <label
          class="position__label"
          for="positionFr-place">
          {{ $t("placeFr") }}
        </label>
        <input
          id="positionFr-place"
          v-model="placeFr"
          class="position__input position__place"
          placeholder="Paris">
      </span>
      <span class="form-group">
        <label
          class="position__label"
          for="positionFr-time">
          {{ $t("timeFr") }}
        </label>
        <input
          id="positionFr-time"
          v-model="timeFr"
          class="position__input position__time"
          placeholder="le 1er mai 2018">
      </span>
    </span>
    <span class="position__language">
      <p class="positionEn__text">
        {{ $t("lastPositionEn") }}
      </p>
      <span class="form-group">
        <label
          class="position__label"
          for="positionEn-place">
          {{ $t("placeEn") }}
        </label>
        <input
          id="positionEn-place"
          v-model="placeEn"
          class="position__input position__place"
          placeholder="London">
      </span>
      <span class="form-group">
        <label
          class="position__label"
          for="positionEn-time">
          {{ $t("timeEn") }}
        </label>
        <input
          id="positionEn-time"
          v-model="timeEn"
          class="position__input position__time"
          placeholder="29th August 2020">
      </span>
    </span>
    <button
      type="submit"
      class="position__button position__action--send"
      @click.prevent="updateLastPosition">
      {{ $t("confirm") }}
    </button>
  </form>
</template>
<script>
  import positionsApi from '../../../services/api/positions'
  import notificationsService from '../../../services/services/notifications'

  export default {
    name: 'PositionForm',
    data() {
      return {
        placeFr: null,
        placeEn: null,
        timeFr: null,
        timeEn: null,
      }
    },
    methods: {
      submit(e) {
        e.preventDefault()
        this.updateLastPosition()
      },
      updateLastPosition() {
        if (this.placeFr === null || this.timeFr === null || this.placeEn === null || this.timeEn === null) {
          this.displayErrorMessage()
        } else {
          const position = {
            place: this.placeFr,
            time: this.timeFr,
            placeEn: this.placeEn,
            timeEn: this.timeEn,
          }
          positionsApi.add(position)
            .then(this.updateLastPositionData)
            .then(this.displaySuccessMessage)
        }
      },
      updateLastPositionData() {
        this.$emit('updateLastPositionData', { place: this.placeFr, time: this.timeFr })
        this.resetPosition()
      },
      resetPosition() {
        this.placeFr = null
        this.placeEn = null
        this.timeFr = null
        this.timeEn = null
      },
      displayErrorMessage() {
        notificationsService.error(this, this.$t('positionNotUpdated'))
      },
      displaySuccessMessage() {
        notificationsService.success(this, this.$t('positionUpdated'))
      },
    },

    i18n: {
      messages: {
        fr: {
          placeFr: 'Position:',
          placeEn: 'Position:',
          timeFr: 'Date:',
          timeEn: 'Date:',
          lastPositionFr: 'Nouvelle position FR :',
          lastPositionEn: 'New position EN :',
          confirm: 'Envoyer',
          positionUpdated: 'Position mise-à-jour',
          positionNotUpdated: 'Renseigne toutes les informations. La position n\'a pas été mise-à-jour',
        },
        en: {
          placeFr: 'Position FR :',
          placeEn: 'Position EN:',
          timeFr: 'Date FR :',
          timeEn: 'Date EN:',
          lastPositionFr: 'Nouvelle position FR:',
          lastPositionEn: 'New position EN:',
          confirm: 'Confirm',
          positionUpdated: 'Position updated',
          positionNotUpdated: 'Fill all the informations. Position was not updated',
        },
      },
    },
  }
</script>
<style scoped>
  .position__form {
    display: flex;
    flex-direction: column;
    align-items: center;
  }


  .form-group {
    text-align: right;
    margin-right: 30px;
    float: right;
  }

  .position__input {
    background: #FFFFFF;
    border: 1px solid cadetblue;
    padding: 15px 10px;
    border-radius: 4px;
    width: 230px;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .position__button {
    text-transform: uppercase;
    color: #f76252;
    background: #FFFFFF;
    border: 1px solid #D14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 230px;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .position__button:hover {
    background: #D14800;
    color: #FFFFFF;
  }

  .position__button:disabled {
    background: #BDBDBD;
    border-color: #616161;
    color: #FAFAFA;
    cursor: auto;
  }

  @media only screen and (min-width: 640px) {
    .form-group {
      text-align: left;
      float: left;
    }
  }

  @media only screen and (min-width: 540px) and (max-width: 992px) {
    .position__form {
      margin-bottom: 60px;
    }
  }
</style>
