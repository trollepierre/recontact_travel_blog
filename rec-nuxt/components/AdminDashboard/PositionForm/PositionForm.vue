<template>
  <form class="position__form">
    <p class="position__text">
      {{ $t("lastPosition") }}
    </p>
    <span class="form-group">
      <label
        class="position__label"
        for="position-place">
        {{ $t("place") }}
      </label>
      <input
        id="position-place"
        v-model="place"
        class="position__input position__place"
        placeholder="Paris">
    </span>
    <span class="form-group">
      <label
        class="position__label"
        for="position-time">
        {{ $t("time") }}
      </label>
      <input
        id="position-time"
        v-model="time"
        class="position__input position__time"
        placeholder="le 1er mai 2018">
    </span>
    <button
      type="submit"
      class="position__button position__action--send"
      @click.prevent="updateLastPosition">
      {{ $t("confirm") }}
    </button>
  </form>
</template>
<script>
  import positionsApi from '../../../services/api/positions'
  import notificationsService from '../../../services/services/notifications'

  export default {
    name: 'PositionForm',
    data() {
      return {
        place: null,
        time: null,
      }
    },
    methods: {
      submit(e) {
        e.preventDefault()
        this.updateLastPosition()
      },
      updateLastPosition() {
        if (this.place === null || this.time === null) {
          this.displayErrorMessage()
        } else {
          const position = {
            place: this.place,
            time: this.time,
          }
          positionsApi.add(position)
            .then(this.updateLastPositionData)
            .then(this.displaySuccessMessage)
        }
      },
      updateLastPositionData() {
        this.$emit('updateLastPositionData', { place: this.place, time: this.time })
        this.resetPosition()
      },
      resetPosition() {
        this.place = null
        this.time = null
      },
      displayErrorMessage() {
        notificationsService.error(this, this.$t('positionNotUpdated'))
      },
      displaySuccessMessage() {
        notificationsService.success(this, this.$t('positionUpdated'))
      },
    },

    i18n: {
      messages: {
        fr: {
          place: 'Position  :',
          time: 'Date  :',
          lastPosition: 'Nouvelle position  :',
          confirm: 'Envoyer',
          lastKnownPosition: 'Dernière position connue  :',
          positionUpdated: 'Position mise-à-jour',
          positionNotUpdated: 'Renseigne toutes les informations. La position n\'a pas été mise-à-jour',
        },
        en: {
          place: 'Position:',
          time: 'Date:',
          lastPosition: 'New position:',
          confirm: 'Confirm',
          lastKnownPosition: 'Last known position:',
          positionUpdated: 'Position updated',
          positionNotUpdated: 'Fill all the informations. Position was not updated',
        },
      },
    },
  }
</script>
<style scoped>
  .form-group {
    text-align: right;
    margin-right: 30px;
    float: right;
  }

  .position__input {
    background: #ffffff;
    border: 1px solid cadetblue;
    padding: 15px 10px;
    border-radius: 4px;
    width: 230px;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .position__button {
    text-transform: uppercase;
    color: #f76252;
    background: #ffffff;
    border: 1px solid #d14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 230px;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .position__button:hover {
    background: #d14800;
    color: #ffffff;
  }

  .position__button:disabled {
    background: #BDBDBD;
    border-color: #616161;
    color: #FAFAFA;
    cursor: auto;
  }

  @media only screen and (min-width: 640px) {
    .form-group {
      text-align: left;
      float: left;
    }
  }

  @media only screen and (min-width: 540px) and (max-width: 992px){
    .position__form {
      margin-bottom: 60px;
    }
  }
</style>
