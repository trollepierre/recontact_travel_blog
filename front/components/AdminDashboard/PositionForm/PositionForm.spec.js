import Vue from 'vue'
import VueI18n from 'vue-i18n'
import PositionForm from './PositionForm.vue'
import positionsApi from '../../../services/api/positions'
import notificationsService from '../../../services/services/notifications'

describe('Component | PositionForm.vue', () => {
  let localVue
  let wrapper

  beforeEach(() => {
    console.warn = jest.fn()
    localVue = createLocalVue()
    localVue.use(VueI18n)
  })

  it('should be named "PositionForm"', () => {
    wrapper = shallowMount(PositionForm, { localVue })

    expect(wrapper.name()).toEqual('PositionForm')
  })

  describe('template', () => {
    it('should match snapshot', () => {
      wrapper = shallowMount(PositionForm, { localVue })

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    let currentPosition
    let currentPositionForApi

    beforeEach(() => {
      wrapper = shallowMount(PositionForm, { localVue })
      currentPosition = {
        placeFr: 'placeFr',
        timeFr: 'timeFr',
        placeEn: 'placeEn',
        timeEn: 'timeEn',
      }
      currentPositionForApi = {
        place: 'placeFr',
        time: 'timeFr',
        placeEn: 'placeEn',
        timeEn: 'timeEn',
      }
      wrapper.setData(currentPosition)
      positionsApi.add = jest.fn()
      positionsApi.add.mockResolvedValue()
      notificationsService.information = jest.fn()
      notificationsService.error = jest.fn()
    })

    describe('#submit', () => {
      it('should prevent default', () => {
        const event = { preventDefault: jest.fn() }

        wrapper.vm.submit(event)

        expect(event.preventDefault).toHaveBeenCalledWith()
      })

      it('should update last position', () => {
        const event = { preventDefault: jest.fn() }

        wrapper.vm.submit(event)

        expect(positionsApi.add).toHaveBeenCalledWith({
          place: 'placeFr',
          time: 'timeFr',
          placeEn: 'placeEn',
          timeEn: 'timeEn',
        })
      })
    })

    describe('#updateLastPosition', () => {
      describe('when position is null', () => {
        beforeEach(() => {
          currentPosition = { placeFr: null, timeFr: 'time' }
          wrapper.setData(currentPosition)
        })

        it('should not add position', () => {
          wrapper.vm.updateLastPosition()

          expect(positionsApi.add).not.toHaveBeenCalled()
        })

        it('should send error notification', () => {
          wrapper.vm.updateLastPosition()

          expect(notificationsService.error).toHaveBeenCalledWith(
            'Fill all the informations. Position was not updated',
          )
        })
      })

      describe('when time is null', () => {
        beforeEach(() => {
          currentPosition = { placeFr: 'place', timeFr: null }
          wrapper.setData(currentPosition)
        })

        it('should not add position', () => {
          wrapper.vm.updateLastPosition()

          expect(positionsApi.add).not.toHaveBeenCalled()
        })

        it('should send error notification', () => {
          wrapper.vm.updateLastPosition()

          expect(notificationsService.error).toHaveBeenCalledWith(
            'Fill all the informations. Position was not updated',
          )
        })
      })

      it('should add current position to api', () => {
        wrapper.vm.updateLastPosition()

        expect(positionsApi.add).toHaveBeenCalledWith(currentPositionForApi)
      })

      xit('should emit current position', async () => {
        await wrapper.vm.updateLastPosition()

        expect(wrapper).toEmit('updateLastPositionData', currentPosition)
      })

      xit('should reset position', async () => {
        await wrapper.vm.updateLastPosition()

        expect(wrapper.vm.place).toEqual(null)
        expect(wrapper.vm.time).toEqual(null)
      })

      it('should send success notifications', async () => {
        notificationsService.information.mockResolvedValue()

        await wrapper.vm.updateLastPosition()

        return Vue.nextTick().then(() => {
          expect(notificationsService.information).toHaveBeenCalledWith(
            'Position updated',
          )
        })
      })
    })

    describe('#updateLastPositionData', () => {
      it('should emit last position', () => {
        wrapper.vm.updateLastPositionData()

        expect(wrapper).toEmit('updateLastPositionData', {
          place: 'placeFr',
          time: 'timeFr',
        })
      })

      it('should reset position', () => {
        wrapper.vm.updateLastPositionData()

        expect(wrapper.vm.placeFr).toEqual(null)
        expect(wrapper.vm.timeFr).toEqual(null)
        expect(wrapper.vm.placeEn).toEqual(null)
        expect(wrapper.vm.timeEn).toEqual(null)
      })
    })

    describe('#resetPosition', () => {
      it('should set position values to null', () => {
        wrapper.vm.resetPosition()

        expect(wrapper.vm.placeFr).toEqual(null)
        expect(wrapper.vm.timeFr).toEqual(null)
        expect(wrapper.vm.placeEn).toEqual(null)
        expect(wrapper.vm.timeEn).toEqual(null)
      })
    })
  })

  describe('events', () => {
    describe('clicking on button "Envoyer"', () => {
      it('should call updateLastPosition', () => {
        wrapper = shallowMount(PositionForm, { localVue })
        wrapper.vm.updateLastPosition = jest.fn()

        wrapper.find('button').trigger('click')

        expect(wrapper.vm.updateLastPosition).toHaveBeenCalled()
      })
    })
  })

  describe('locales', () => {
    const languages = Object.keys(PositionForm.i18n.messages)

    it('contains 2 languages', () => {
      expect(languages).toHaveLength(2)
      expect(languages).toEqual(['fr', 'en'])
    })

    describe('each language', () => {
      describe('fr', () => {
        const locales = Object.keys(PositionForm.i18n.messages.fr)

        it('contains 9 locales', () => {
          expect(locales).toHaveLength(9)
          expect(locales).toMatchInlineSnapshot(`
            Array [
              "placeFr",
              "placeEn",
              "timeFr",
              "timeEn",
              "lastPositionFr",
              "lastPositionEn",
              "confirm",
              "positionUpdated",
              "positionNotUpdated",
            ]
          `)
        })
      })

      describe('en', () => {
        const locales = Object.keys(PositionForm.i18n.messages.en)

        it('contains 9 locales', () => {
          expect(locales).toHaveLength(9)
          expect(locales).toMatchInlineSnapshot(`
            Array [
              "placeFr",
              "placeEn",
              "timeFr",
              "timeEn",
              "lastPositionFr",
              "lastPositionEn",
              "confirm",
              "positionUpdated",
              "positionNotUpdated",
            ]
          `)
        })
      })
    })
  })
})
