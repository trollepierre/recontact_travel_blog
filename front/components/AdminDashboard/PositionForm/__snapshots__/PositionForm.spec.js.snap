// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | PositionForm.vue template should match snapshot 1`] = `
<form
  class="position__form"
>
  <p
    class="position__text"
  >
    
    lastPosition
  
  </p>
   
  <span
    class="form-group"
  >
    <label
      class="position__label"
      for="position-place"
    >
      
      place
    
    </label>
     
    <input
      class="position__input position__place"
      id="position-place"
      placeholder="Paris"
    />
  </span>
   
  <span
    class="form-group"
  >
    <label
      class="position__label"
      for="position-time"
    >
      
      time
    
    </label>
     
    <input
      class="position__input position__time"
      id="position-time"
      placeholder="le 1er mai 2018"
    />
  </span>
   
  <button
    class="position__button position__action--send"
    type="submit"
  >
    
    confirm
  
  </button>
</form>
`;
